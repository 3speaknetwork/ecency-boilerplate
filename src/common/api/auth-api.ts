import axios from "axios";

import {_u} from "./private-api";

export const hsTokenRenew = (code: string): Promise<{
    username: string;
    access_token: string;
    refresh_token: string;
    expires_in: number;
}> =>
    axios
        .post(_u(`/api/hs-token-refresh`), {
            code,
        })
        .then((resp) => resp.data);

