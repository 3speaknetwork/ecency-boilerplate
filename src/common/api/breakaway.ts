import axios from "axios"

// const baUrl = "http://localhost:4000"
const baUrl = "https://breakaway-points-system-api.onrender.com"

export const createBreakawayUser = async (username: string, community: string, referral: string, email: string)=> {
    try {
        const data = {
            username,
            community,
            referral,
            email
        }
        const resp = await axios.post(`${baUrl}/signup-keychain`, data)
        return resp;
    } catch (err) {
        console.log(err)
        return err;
    }
}