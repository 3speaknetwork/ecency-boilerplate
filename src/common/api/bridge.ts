import { Entry } from "../store/entries/types";
import { Community } from "../store/communities/types";
import { Subscription } from "../store/subscriptions/types";

import { Client } from "@hiveio/dhive";
import SERVERS from "../constants/servers.json";
import { client } from "./hive";

export const bridgeServer = new Client(SERVERS, {
  timeout: 3000,
  failoverThreshold: 3,
  consoleOnFailover: true
});
export const dataLimit = typeof window !== "undefined" && window.screen.width < 540 ? 5 : 20 || 20;

const bridgeApiCall = <T>(endpoint: string, params: {}): Promise<T> =>
  bridgeServer.call("bridge", endpoint, params);

const resolvePost = (post: Entry, observer: string): Promise<Entry> => {
  const { json_metadata: json } = post;

  if (
    json.original_author &&
    json.original_permlink &&
    json.tags &&
    json.tags[0] === "cross-post"
  ) {
    return getPost(json.original_author, json.original_permlink, observer)
      .then((resp) => {
        if (resp) {
          return {
            ...post,
            original_entry: resp
          };
        }

        return post;
      })
      .catch(() => {
        return post;
      });
  }

  return new Promise((resolve) => {
    resolve(post);
  });
};

const resolvePosts = (posts: Entry[], observer: string): Promise<Entry[]> => {
  const promises = posts.map((p) => resolvePost(p, observer));

  return Promise.all(promises);
};

export const getPostsRanked = (
  sort: string,
  start_author: string = "",
  start_permlink: string = "",
  limit: number = dataLimit,
  tag: string = "",
  observer: string = ""
): Promise<Entry[] | null> => {
  return bridgeApiCall<Entry[] | null>("get_ranked_posts", {
    sort,
    start_author,
    start_permlink,
    limit,
    tag,
    observer
  }).then((resp) => {
    if (resp) {
      return resolvePosts(resp, observer);
    }

    return resp;
  });
};

export const getAccountPosts = (
  sort: string,
  account: string,
  start_author: string = "",
  start_permlink: string = "",
  limit: number = dataLimit,
  observer: string = ""
): Promise<Entry[] | null> => {
  return bridgeApiCall<Entry[] | null>("get_account_posts", {
    sort,
    account,
    start_author,
    start_permlink,
    limit,
    observer
  }).then((resp) => {
    if (resp) {
      return resolvePosts(resp, observer);
    }

    return resp;
  });
};

export const getPost = (
  author: string = "",
  permlink: string = "",
  observer: string = ""
): Promise<Entry | null> => {
  return bridgeApiCall<Entry | null>("get_post", {
    author,
    permlink,
    observer
  }).then((resp) => {
    if (resp) {
      return resolvePost(resp, observer);
    }

    return resp;
  });
};

export interface AccountNotification {
  date: string;
  id: number;
  msg: string;
  score: number;
  type: string;
  url: string;
}

export const getAccountNotifications = (
  account: string,
  lastId: number | null = null,
  limit = 50
): Promise<AccountNotification[] | null> => {
  const params: { account: string; last_id?: number; limit: number } = {
    account,
    limit
  };

  if (lastId) {
    params.last_id = lastId;
  }

  return bridgeApiCall<AccountNotification[] | null>("account_notifications", params);
};

export const getDiscussion = (
  author: string,
  permlink: string
): Promise<Record<string, Entry> | null> =>
  bridgeApiCall<Record<string, Entry> | null>("get_discussion", {
    author,
    permlink
  });

export const getCommunity = (
  name: string,
  observer: string | undefined = ""
): Promise<Community | null> =>
  bridgeApiCall<Community | null>("get_community", { name, observer });

export const getCommunities = (
  last: string = "",
  limit: number = 100,
  query?: string | null,
  sort: string = "rank",
  observer: string = ""
): Promise<Community[] | null> =>
  bridgeApiCall<Community[] | null>("list_communities", {
    last,
    limit,
    query,
    sort,
    observer
  });

export const normalizePost = (post: any): Promise<Entry | null> =>
  bridgeApiCall<Entry | null>("normalize_post", {
    post
  });

export const getSubscriptions = (account: string): Promise<Subscription[] | null> =>
  bridgeApiCall<Subscription[] | null>("list_all_subscriptions", {
    account
  });

export const getSubscribers = (community: string): Promise<Subscription[] | null> =>
  bridgeApiCall<Subscription[] | null>("list_subscribers", {
    community
  });

export interface AccountRelationship {
  follows: boolean;
  ignores: boolean;
  is_blacklisted: boolean;
  follows_blacklists: boolean;
}

export const getRelationshipBetweenAccounts = (
  follower: string,
  following: string
): Promise<AccountRelationship | null> =>
  bridgeApiCall<AccountRelationship | null>("get_relationship_between_accounts", [
    follower,
    following
  ]);

// @source https://ecency.com/hive-139531/@andablackwidow/rc-stats-in-1-27
export type RcOperation =
  | "comment_operation"
  | "vote_operation"
  | "transfer_operation"
  | "custom_json_operation";

export interface RcOperationStats {
  count: number; // number of such operations executed during last day
  avg_cost_rc: number; // average RC cost of single operation
  resource_cost: {
    // average RC cost split between various resources
    history_rc: number;
    tokens_rc: number;
    market_rc: number;
    state_rc: number;
    exec_rc: number;
  };
  resource_cost_share: {
    // share of resource cost in average final cost (expressed in basis points)
    history_bp: number;
    tokens_bp: number;
    market_bp: number;
    state_bp: number;
    exec_bp: number;
  };
  resource_usage: {
    // average consumption of resources per operation
    history_bytes: number; // - size of transaction in bytes
    tokens: string; // - number of tokens (always 0 or 1 (with exception of multiop) - tokens are internally expressed with 4 digit precision
    market_bytes: number; // - size of transaction in bytes when it belongs to market category or 0 otherwise
    state_hbytes: number; // - hour-bytes of state
    exec_ns: number; // - nanoseconds of execution time
  };
}

export const getRcOperationStats = (): Promise<any> =>
  bridgeServer.call("rc_api", "get_rc_stats", {});
