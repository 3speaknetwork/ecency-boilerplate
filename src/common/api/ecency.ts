import axios from 'axios';

import defaults from "../constants/defaults.json";

export const uploadImage = async (file: File, token: string): Promise<{
    url: string
}> => {
    const fData = new FormData();
    fData.append('file', file);

    const postUrl = `${defaults.imageServer}/hs/${token}`;

    return axios.post(postUrl, fData, {
        headers: {
            'Content-Type': 'multipart/form-data'
        }
    }).then(r => r.data);
};
