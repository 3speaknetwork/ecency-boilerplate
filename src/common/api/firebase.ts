import { getMessaging, getToken, MessagePayload, Messaging, onMessage } from '@firebase/messaging';
import { FirebaseApp, initializeApp } from '@firebase/app';
import { _t } from '../i18n';
import { NotifyTypes } from '../enums';

let app: FirebaseApp;
export let FCM: Messaging;

export function initFirebase() {
  if (typeof window === 'undefined') {
    return;
  }

  app = initializeApp({
    apiKey: 'AIzaSyDKF-JWDMmUs5ozjK7ZdgG4beHRsAMd2Yw',
    authDomain: 'esteem-ded08.firebaseapp.com',
    databaseURL: 'https://esteem-ded08.firebaseio.com',
    projectId: 'esteem-ded08',
    storageBucket: 'esteem-ded08.appspot.com',
    messagingSenderId: '211285790917',
    appId: '1:211285790917:web:c259d25ed1834c683760ac',
    measurementId: 'G-TYQD1N3NR3'
  });
  FCM = getMessaging(app);
}

export const handleMessage = (payload: MessagePayload) => {
  const notificationTitle = payload.notification?.title || 'Ecency';

  const notification = new Notification(notificationTitle, {
    body: payload.notification?.body,
    icon: payload.notification?.image,
  });

  notification.onclick = () => {
    let url = 'https://ecency.com';
    const data = (payload.data || {}) as any;

    if (data.parent_permlink1) {
      url += '/' + data.parent_permlink1;
    }
    if (['vote', 'unvote', 'reply', 'spin', 'inactive'].includes(data.type)) {
      url += '/@' + data.target;
    } else {
      // delegation, mention, transfer, follow, unfollow, ignore, blacklist, reblog
      url += '/@' + data.source;
    }
    if (data.permlink1) {
      url += '/' + data.permlink1;
    }

    window.open(url, '_blank');
  }
}

export const getFcmToken = () => getToken(FCM, {
  vapidKey: 'BA3SrGKAKMU_6PXOFwD9EQ1wIPzyYt90Q9ByWb3CkazBe8Isg7xr9Cgy0ka6SctHDW0VZLShTV_UDYNxewzWDjk'
});

export const listenFCM = (callback: Function) => {
  onMessage(FCM, p => {
    //console.log('Received fg message', p);
    handleMessage(p);
    callback();
  });
}