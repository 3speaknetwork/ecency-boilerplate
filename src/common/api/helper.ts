import isElectron from "../util/is-electron";

import defaults from "../constants/defaults.json";

export const apiBase = (endpoint: string): string => {
    let base = '';

    if (isElectron()) {
        if (process.env.NODE_ENV === "development") {
            base = "http://localhost:3000";
        } else {
            base = defaults.base;
        }
    }

    return `${base}${endpoint}`;
}
