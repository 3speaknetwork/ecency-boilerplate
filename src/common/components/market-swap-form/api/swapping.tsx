import { MarketAsset } from "../market-pair";
import { PrivateKey } from "@hiveio/dhive";
import React from "react";

export enum SwappingMethod {
  KEY = "key",
  HS = "hs",
  KC = "kc",
  CUSTOM = "custom"
}

export const MarketSwappingMethods = {
  [MarketAsset.HIVE]: [SwappingMethod.HS, SwappingMethod.KC, SwappingMethod.KEY],
  [MarketAsset.HBD]: [SwappingMethod.HS, SwappingMethod.KC, SwappingMethod.KEY]
};

export interface SwapOptions {
  fromAsset: MarketAsset;
  toAsset: MarketAsset;
  fromAmount: string;
  toAmount: string;
}

export const swapByKey = (key: PrivateKey, options: SwapOptions) => {};

export const swapByKc = (options: SwapOptions) => {};

export const swapByHs = (options: SwapOptions) => {};
