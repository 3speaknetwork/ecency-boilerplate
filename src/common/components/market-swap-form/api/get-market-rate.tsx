import { MarketAsset } from "../market-pair";
import { getMarketStatistics, getOrderBook, OrdersDataItem } from "../../../api/hive";
import React, { useEffect, useState } from "react";
import { error } from "../../feedback";

export const getMarketRate = async (asset: MarketAsset): Promise<number> => {
  if (asset === MarketAsset.HIVE) {
    const market = await getMarketStatistics();
    return +market.lowest_ask;
  } else if (asset === MarketAsset.HBD) {
    const market = await getMarketStatistics();
    return 1 / +market.lowest_ask;
  }
  return 0;
};

interface Props {
  asset: MarketAsset;
  amount: string;
  setToAmount: (amount: string) => void;
  loading: boolean;
  setLoading: (v: boolean) => void;
  setInvalidAmount: (v: boolean) => void;
  setTooMuchSlippage: (v: boolean) => void;
}

export const MarketRateListener = ({
  asset,
  amount,
  setToAmount,
  setLoading,
  setInvalidAmount,
  setTooMuchSlippage
}: Props) => {
  const [hiveOrderBook, setHiveOrderBook] = useState<OrdersDataItem[]>([]);
  const [hbdOrderBook, setHbdOrderBook] = useState<OrdersDataItem[]>([]);

  useEffect(() => {
    fetchHiveOrderBook().then(() => processHiveOrderBook());
  }, []);

  useEffect(() => {
    fetchHiveOrderBook().then(() => processHiveOrderBook());
  }, [asset]);

  useEffect(() => {
    processHiveOrderBook();
  }, [amount]);

  const calculatePrice = (intAmount: number, book: OrdersDataItem[], asset: "hive" | "hbd") => {
    let available = 0;
    let index = 0;
    while (available < intAmount && book.length > index + 1) {
      available += book[index][asset] / 1000;
      index++;
    }
    return +book[index].real_price;
  };

  const processHiveOrderBook = async () => {
    if (hiveOrderBook.length > 0 && hbdOrderBook.length > 0) {
      const intAmount = +amount.replace(/,/gm, "");

      let availableInOrderBook = 0;
      let firstPrice = Infinity;
      let price = 0;
      let toAmount = "";

      if (asset === MarketAsset.HIVE) {
        availableInOrderBook =
          hbdOrderBook.map((item) => item.hive).reduce((acc, item) => acc + item, 0) / 1000;
        price = calculatePrice(intAmount, hbdOrderBook, "hive");
        toAmount = intAmount * price + "";
        firstPrice = +hbdOrderBook[0].real_price;
      } else if (asset === MarketAsset.HBD) {
        availableInOrderBook =
          hiveOrderBook.map((item) => item.hbd).reduce((acc, item) => acc + item, 0) / 1000;
        price = calculatePrice(intAmount, hiveOrderBook, "hbd");
        toAmount = intAmount / price + "";
        firstPrice = +hiveOrderBook[0].real_price;
      }

      if (!availableInOrderBook) return null;

      const slippage = Math.abs(price - firstPrice);
      setTooMuchSlippage(slippage > 0.001);

      if (intAmount > availableInOrderBook) {
        setInvalidAmount(true);
      } else if (toAmount) {
        setToAmount(toAmount);
        setInvalidAmount(false);
      }
    }

    return null;
  };

  const fetchHiveOrderBook = async () => {
    setLoading(true);
    try {
      const book = await getOrderBook();
      setHiveOrderBook(book.bids);
      setHbdOrderBook(book.asks);
    } catch (e) {
      error("Order book is empty.");
    } finally {
      setLoading(false);
    }
    return null;
  };

  return <></>;
};
