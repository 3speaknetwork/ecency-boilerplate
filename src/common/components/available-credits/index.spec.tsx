import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import { AvailableCredits } from "./index";
import { RcOperation } from "../../api/bridge";

it("(1) Default render", () => {
  const props = {
    username: "test",
    operation: "comment_operation" as RcOperation
  };
  const component = renderer.create(
    <BrowserRouter>
      <AvailableCredits {...props} />
    </BrowserRouter>
  );
  expect(component.toJSON()).toMatchSnapshot();
});
