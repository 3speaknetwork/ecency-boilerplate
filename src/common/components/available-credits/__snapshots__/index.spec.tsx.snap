// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(1) Default render 1`] = `
Array [
  <div
    className="available-credits d-flex align-items-center justify-content-between w-100 pr-2"
  >
    <div
      className="available-credits-bar w-100 undefined danger"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
    >
      <div
        className="progress"
      >
        <div
          className="indicator danger"
          style={
            Object {
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="btn btn-link px-0"
      onClick={[Function]}
    >
      Boost
    </div>
  </div>,
  <div
    className="available-credits-bar-popper "
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
  >
    <div>
      <div
        className="p-2"
      >
        <span
          className="opacity-75"
        >
          Resource credits
        </span>
        <div>
          0
          (
          0
          %)
        </div>
        <div>
          <small>
            (fully charged a few seconds ago)
          </small>
        </div>
      </div>
      <div
        className="delegations d-flex flex-column p-2"
      >
        <span
          className="incoming mb-1"
        >
          <div
            className="opacity-75"
          >
            Delegations in
          </div>
          0
        </span>
        <span
          className="outgoing"
        >
          <div
            className="opacity-75"
          >
            Delegations out
          </div>
          0
        </span>
      </div>
    </div>
    <div
      className="extra-details p-2"
    >
      <span
        className="d-block mb-2 opacity-5"
      >
        What you can do with RC?
      </span>
      <div
        className="extras"
      >
        <div
          className="mb-1"
        >
          <div
            className="opacity-75"
          >
            Posts/Comments
          </div>
          0
        </div>
        <div
          className="two-col"
        >
          <div
            className="mb-1"
          >
            <div
              className="opacity-75"
            >
              Votes
            </div>
            0
          </div>
          <div>
            <div
              className="opacity-75"
            >
              Transfers
            </div>
            0
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
